Automattic/jetpack
# exploring a menu registration package

This is an exploratory PR with a proposal on how a package to handle menu registration can look like. If we agree one this, I'll finish it up and break this PR in smaller pieces. (First adding the package and then using it in the plugins)

I spent quite some time going through all the code that handles menu registration in the jetpack plugin. We have the `Jetpack_Admin_Page` class and a bunch of logic there and many menu items are registered inside the individual modules.

At first I thought about refactoring it a bit, in order to use this package I was about to create, but I finally thought the best thing was not to touch it at all. At first, this package will be useful for the stand-alone plugins and eventually we can modify the Jetpack plugin to use it as well.

So the outcome was a very minimalistic package that will basically make sure that pages are always added under the Jetpack top level menu - and will register the top level menu if it's not there yet.

It will:

* give a simple method for stand-alone plugins to add a new menu item
* Check if Jetpack top level menu is already registered, and will register it if not
* If Jetpack plugin is not there, it will also enqueue the font and make sure the icon is properly added to the menu
* Check user capability needed to see the menu item and hide the Jetpack menu altogether if the user does not have enough capabilities to see at least one of the registered sub-menus

How to test:
* In this PR, Backup plugin is using the package. Alternate having the Backup plugin activated while Jetpack plugin is both active and inactive.
* Change the required capability to see the Backup menu item and/or test it with different user roles. (`class-jetpack-backup.php` line 31)

Things to check:
* Does this approach looks right?
* Is this a good package name? or should we go to something more specific?
* Should we include the fonts, duplicating them from the plugin? (we can remove them from the plugin later and make it use this package)
* Does the capability check work well and hide the whole top level menu if the user can't see anything?
